File: applicatie/src/main/java/com/vdzon/administratie/pdfgenerator/pdfutil/GeneratePdfHelper.java
package com.vdzon.administratie.pdfgenerator.pdfutil;

import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.pdmodel.font.PDFont;
import org.apache.pdfbox.pdmodel.graphics.image.PDImageXObject;

import java.io.*;
import java.net.URL;

public class GeneratePdfHelper {
    public static final int NORMAL_TEXT_SKIPDOWN = 15;
    public static final int TITLE_TEXT_SKIPDOWN = 20;
    public static final int TITLE_FONT_SIZE = 19;
    public static final int NORMAL_FONT_SIZE = 12;
    public static final int SKIPDOWN_LINE = 5;
    public static final int TEXT_POS_LEFT = 30;
    private PdfData pdfData;

    public GeneratePdfHelper(PdfData pdfData) {
        this.pdfData = pdfData;
    }

    public void addAnImage(PDDocument document, String logoUrl) {
        try {
            saveLogoToFile(logoUrl, "logo.png");
            PDImageXObject ximage = PDImageXObject.createFromFile("logo.png", document);
            // TODO: waarom scale? Kan ik hier geen vast height gebruiken?
            float scale = 0.5f;
            skipDown( ximage.getHeight() * scale);
            pdfData.page.drawImage(ximage, 30, pdfData.pos, ximage.getWidth() * scale, ximage.getHeight() * scale);
        } catch (IOException ioex) {
            System.out.println("No image for you");
        }
    }

    private void saveLogoToFile(String logoUrl, String filename) throws IOException {
        URL url = new URL(logoUrl);
        InputStream in = new BufferedInputStream(url.openStream());
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        byte[] buf = new byte[1024];
        int n = 0;
        while (-1 != (n = in.read(buf))) {
            out.write(buf, 0, n);
        }
        out.close();
        in.close();
        byte[] response = out.toByteArray();
        FileOutputStream fos = new FileOutputStream(filename);
        fos.write(response);
        fos.close();
    }

    public void writeNormalText(String text) throws IOException {
        skipDown(NORMAL_TEXT_SKIPDOWN);
        writeText( NORMAL_FONT_SIZE, pdfData.fontPlain, text);
    }

    public void writeBoldText(String text) throws IOException {
        skipDown(NORMAL_TEXT_SKIPDOWN);
        writeText(NORMAL_FONT_SIZE, pdfData.fontBold, text);
    }

    public void writeTitle(String text) throws IOException {
        skipDown(TITLE_TEXT_SKIPDOWN);
        writeText(TITLE_FONT_SIZE, pdfData.fontBold, text);
    }

    public void writeText(float fontSize, PDFont fontPlain, String text) throws IOException {
        writeText( fontSize, TEXT_POS_LEFT, fontPlain, text);
    }

    public void writeText(float fontSize, float x, PDFont fontPlain, String text) throws IOException {
        if (text == null) return;
        pdfData.page.beginText();
        pdfData.page.setFont(fontPlain, fontSize);
        pdfData.page.newLineAtOffset(x, pdfData.pos);
        pdfData.page.showText(text);
        pdfData.page.endText();
    }

    public void writeTabel5(PDFont font, TabelCols cols, String titel1, String titel2, String titel3, String titel4, String titel5) throws IOException {
        skipDown(NORMAL_TEXT_SKIPDOWN);
        writeText(NORMAL_FONT_SIZE, cols.getCol(0), font, titel1);
        writeText(NORMAL_FONT_SIZE, cols.getCol(1), font, titel2);
        writeText(NORMAL_FONT_SIZE, cols.getCol(2), font, titel3);
        writeText(NORMAL_FONT_SIZE, cols.getCol(3), font, titel4);
        writeText(NORMAL_FONT_SIZE, cols.getCol(4), font, titel5);
    }

    public void writeTabel3(PDFont font, TabelCols cols, String titel1, String titel2, String titel3) throws IOException {
        skipDown(NORMAL_TEXT_SKIPDOWN);
        writeText(NORMAL_FONT_SIZE, cols.getCol(0), font, titel1);
        writeText(NORMAL_FONT_SIZE, cols.getCol(1), font, titel2);
        writeText(NORMAL_FONT_SIZE, cols.getCol(2), font, titel3);
    }

    public float skipDown(float amount) {
        return pdfData.pos -= amount;
    }

    public void drawLine() throws IOException {
        skipDown(SKIPDOWN_LINE);
        pdfData.page.setLineWidth(1);
        pdfData.page.moveTo(0, pdfData.pos);
        pdfData.page.lineTo(pdfData.pageWidth, pdfData.pos);
        pdfData.page.closeAndStroke();
    }
}

File: pdfgenerator/impl/src/main/java/com/vdzon/administratie/pdfgenerator/pdfutil/GeneratePdfHelper.kt
package com.vdzon.administratie.pdfgenerator.pdfutil

import org.apache.pdfbox.pdmodel.PDDocument
import org.apache.pdfbox.pdmodel.font.PDFont
import org.apache.pdfbox.pdmodel.graphics.image.PDImageXObject

import java.io.*
import java.net.URL

class GeneratePdfHelper(private val pdfData: PdfData) {

    fun addAnImage(document: PDDocument, logoUrl: String) {
        try {
            saveLogoToFile(logoUrl, "logo.png")
            val ximage = PDImageXObject.createFromFile("logo.png", document)
            // TODO: waarom scale? Kan ik hier geen vast height gebruiken?
            val scale = 0.5f
            skipDown(ximage.height * scale)
            pdfData.page!!.drawImage(ximage, 30f, pdfData.pos, ximage.width * scale, ximage.height * scale)
        } catch (ioex: IOException) {
            println("No image for you")
        }

    }

    @Throws(IOException::class)
    private fun saveLogoToFile(logoUrl: String, filename: String) {
        val url = URL(logoUrl)
        val bufIn = BufferedInputStream(url.openStream())
        val out = ByteArrayOutputStream()
        val buf = ByteArray(1024)
        var n = 0
        while ({n = bufIn.read(buf);n}() != -1) {
            out.write(buf, 0, n)
        }

        out.close()
        bufIn.close()
        val response = out.toByteArray()
        val fos = FileOutputStream(filename)
        fos.write(response)
        fos.close()
    }

    @Throws(IOException::class)
    fun writeNormalText(text: String) {
        skipDown(NORMAL_TEXT_SKIPDOWN.toFloat())
        writeText(NORMAL_FONT_SIZE.toFloat(), pdfData.fontPlain, text)
    }

    @Throws(IOException::class)
    fun writeBoldText(text: String) {
        skipDown(NORMAL_TEXT_SKIPDOWN.toFloat())
        writeText(NORMAL_FONT_SIZE.toFloat(), pdfData.fontBold, text)
    }

    @Throws(IOException::class)
    fun writeTitle(text: String) {
        skipDown(TITLE_TEXT_SKIPDOWN.toFloat())
        writeText(TITLE_FONT_SIZE.toFloat(), pdfData.fontBold, text)
    }

    @Throws(IOException::class)
    fun writeText(fontSize: Float, fontPlain: PDFont, text: String) {
        writeText(fontSize, TEXT_POS_LEFT.toFloat(), fontPlain, text)
    }

    @Throws(IOException::class)
    fun writeText(fontSize: Float, x: Float, fontPlain: PDFont, text: String?) {
        if (text == null) return
        pdfData.page!!.beginText()
        pdfData.page!!.setFont(fontPlain, fontSize)
        pdfData.page!!.newLineAtOffset(x, pdfData.pos)
        pdfData.page!!.showText(text)
        pdfData.page!!.endText()
    }

    @Throws(IOException::class)
    fun writeTabel5(font: PDFont, cols: TabelCols, titel1: String, titel2: String, titel3: String, titel4: String, titel5: String) {
        skipDown(NORMAL_TEXT_SKIPDOWN.toFloat())
        writeText(NORMAL_FONT_SIZE.toFloat(), cols.getCol(0).toFloat(), font, titel1)
        writeText(NORMAL_FONT_SIZE.toFloat(), cols.getCol(1).toFloat(), font, titel2)
        writeText(NORMAL_FONT_SIZE.toFloat(), cols.getCol(2).toFloat(), font, titel3)
        writeText(NORMAL_FONT_SIZE.toFloat(), cols.getCol(3).toFloat(), font, titel4)
        writeText(NORMAL_FONT_SIZE.toFloat(), cols.getCol(4).toFloat(), font, titel5)
    }

    @Throws(IOException::class)
    fun writeTabel3(font: PDFont, cols: TabelCols, titel1: String, titel2: String, titel3: String) {
        skipDown(NORMAL_TEXT_SKIPDOWN.toFloat())
        writeText(NORMAL_FONT_SIZE.toFloat(), cols.getCol(0).toFloat(), font, titel1)
        writeText(NORMAL_FONT_SIZE.toFloat(), cols.getCol(1).toFloat(), font, titel2)
        writeText(NORMAL_FONT_SIZE.toFloat(), cols.getCol(2).toFloat(), font, titel3)
    }

    fun skipDown(amount: Float): Float {
        pdfData.pos -= amount
        return pdfData.pos
    }

    @Throws(IOException::class)
    fun drawLine() {
        skipDown(SKIPDOWN_LINE.toFloat())
        pdfData.page!!.setLineWidth(1f)
        pdfData.page!!.moveTo(0f, pdfData.pos)
        pdfData.page!!.lineTo(pdfData.pageWidth, pdfData.pos)
        pdfData.page!!.closeAndStroke()
    }

    companion object {
        val NORMAL_TEXT_SKIPDOWN = 15
        val TITLE_TEXT_SKIPDOWN = 20
        val TITLE_FONT_SIZE = 19
        val NORMAL_FONT_SIZE = 12
        val SKIPDOWN_LINE = 5
        val TEXT_POS_LEFT = 30
    }
}